### ISSUE TITLE
Error when compiling with CUDA on Windows

### ISSUE THREAD

**Issue by BennyH26** [withcatai/node-llama-cpp#40]
### Issue description

On Windows 10, when calling `node-llama-cpp download --cuda` I receive an `nvcc` compiler error

### Expected Behavior

I expect the operation to complete successfully

### Actual Behavior

The actual output of this command is below. Of note, `nvcc fatal   : A single input file is required for a non-link phase when an outputfile is specified`

```
Repo: ggerganov/llama.cpp
Release: b1204
CUDA: enabled

√ Fetched llama.cpp info
√ Removed existing llama.cpp directory
Cloning llama.cpp
Clone ggerganov/llama.cpp  100% ████████████████████████████████████████  0s
◷ Compiling llama.cpp
Not searching for unused variables given on the command line.
-- Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.29.30151.0
-- The CXX compiler identification is MSVC 19.29.30151.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: C:/Program Files/Git/cmd/git.exe (found version "2.40.1.windows.1")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE
-- Found CUDAToolkit: Y:/CUDA_12/include (found version "12.2.140")
-- cuBLAS found
-- The CUDA compiler identification is NVIDIA 12.2.140
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: Y:/CUDA_12/bin/nvcc.exe - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Using CUDA architectures: 52;61;70
-- CMAKE_SYSTEM_PROCESSOR: AMD64
-- x86 detected
-- Configuring done (15.7s)
-- Generating done (0.1s)
-- Build files have been written to: C:/Users/Ben/Development/test_app/node_modules/node-llama-cpp/llama/build
Microsoft (R) Build Engine version 16.11.2+f32259642 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Generating build details from Git
  -- Found Git: C:/Program Files/Git/cmd/git.exe (found version "2.40.1.windows.1")
  Building Custom Rule C:/Users/Ben/Development/test_app/node_modules/node-llama-cpp/llama/llama.cpp/CMakeLists.txt
  Building Custom Rule C:/Users/Ben/Development/test_app/node_modules/node-llama-cpp/llama/llama.cpp/CMakeLists.txt
  Compiling CUDA source file ..\..\llama.cpp\ggml-cuda.cu...

  C:\Users\User\Development\test_app\node_modules\node-llama-cpp\llama\build\llama.cpp>"Y:\CUDA_12\bin\nvcc.exe"  --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64" -x cu   -I"C:\Users\User\Devel
  opment\test_app\node_modules\node-addon-api" -I"C:\Users\User\.cmake-js\node-x64\v19.9.0\include\node" -I"C:\Users\User\Development\test_app\node_modules\node-llama-cpp\llama\llama.cpp\." -IY:\CUDA_12\include -IY:\CUDA_12\include     --keep-dir x64\Release  -maxrregcount=0
    --machine 64 --compile -cudart static --generate-code=arch=compute_52,code=[compute_52,sm_52] --generate-code=arch=compute_61,code=[compute_61,sm_61] --generate-code=arch=compute_70,code=[compute_70,sm_70] /EHsc -Xcompiler="/EHsc -Ob2"   -D_WINDOWS -DNDEBUG -DNAPI_VERSION=7 -DGGML_USE_K_
  QUANTS -DGGML_USE_CUBLAS -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -D_CRT_SECURE_NO_WARNINGS -D_XOPEN_SOURCE=600 -D"CMAKE_INTDIR=\"Release\"" -D_MBCS -DWIN32 -D_WINDOWS -DNDEBUG -DNAPI_VERSION=7 -DGGML_USE_K_QUANTS -DGGML_USE_CUBLAS -DGGML_CUDA_DMMV_X=32 -DGGML_C
  UDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -D_CRT_SECURE_NO_WARNINGS -D_XOPEN_SOURCE=600 -D"CMAKE_INTDIR=\"Release\"" -Xcompiler "/EHsc /W3 /nologo /O2 /FS   /MD /GR" -Xcompiler "/Fdggml.dir\Release\ggml.pdb" -o ggml.dir\Release\ggml-cuda.obj "C:\Users\User\Development\test_app\nod
  e_modules\node-llama-cpp\llama\llama.cpp\ggml-cuda.cu"
  nvcc fatal   : A single input file is required for a non-link phase when an outputfile is specified
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\BuildCustomizations\CUDA 12.2.targets(799,9): error MSB3721: The command ""Y:\CUDA_12\bin\nvcc.exe"  --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.
29.30133\bin\HostX64\x64" -x cu   -I"C:\Users\User\Development\test_app\node_modules\node-addon-api" -I"C:\Users\User\.cmake-js\node-x64\v19.9.0\include\node" -I"C:\Users\User\Development\test_app\node_modules\node-llama-cpp\llama\llama.cpp\." -IY:\CUDA_12\include -IY:
\CUDA_12\include     --keep-dir x64\Release  -maxrregcount=0   --machine 64 --compile -cudart static --generate-code=arch=compute_52,code=[compute_52,sm_52] --generate-code=arch=compute_61,code=[compute_61,sm_61] --generate-code=arch=compute_70,code=[compute_70,sm_70] /EHsc -Xcompiler="/EHsc
 -Ob2"   -D_WINDOWS -DNDEBUG -DNAPI_VERSION=7 -DGGML_USE_K_QUANTS -DGGML_USE_CUBLAS -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -D_CRT_SECURE_NO_WARNINGS -D_XOPEN_SOURCE=600 -D"CMAKE_INTDIR=\"Release\"" -D_MBCS -DWIN32 -D_WINDOWS -DNDEBUG -DNAPI_VERSION=7 -DGGML_USE_
K_QUANTS -DGGML_USE_CUBLAS -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -D_CRT_SECURE_NO_WARNINGS -D_XOPEN_SOURCE=600 -D"CMAKE_INTDIR=\"Release\"" -Xcompiler "/EHsc /W3 /nologo /O2 /FS   /MD /GR" -Xcompiler "/Fdggml.dir\Release\ggml.pdb" -o ggml.dir\Release\ggml-cuda.
obj "C:\Users\User\Development\test_app\node_modules\node-llama-cpp\llama\llama.cpp\ggml-cuda.cu"" exited with code 1. [C:\Users\User\Development\test_app\node_modules\node-llama-cpp\llama\build\llama.cpp\ggml.vcxproj]
Not searching for unused variables given on the command line.
-- Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.29.30151.0
-- The CXX compiler identification is MSVC 19.29.30151.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: C:/Program Files/Git/cmd/git.exe (found version "2.40.1.windows.1")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE
-- Found CUDAToolkit: Y:/CUDA_12/include (found version "12.2.140")
-- cuBLAS found
-- The CUDA compiler identification is NVIDIA 12.2.140
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: Y:/CUDA_12/bin/nvcc.exe - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Using CUDA architectures: 52;61;70
-- CMAKE_SYSTEM_PROCESSOR: AMD64
-- x86 detected
-- Configuring done (15.9s)
-- Generating done (0.1s)
-- Build files have been written to: C:/Users/Ben/Development/test_app/node_modules/node-llama-cpp/llama/build
Microsoft (R) Build Engine version 16.11.2+f32259642 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Generating build details from Git
  -- Found Git: C:/Program Files/Git/cmd/git.exe (found version "2.40.1.windows.1")
  Building Custom Rule C:/Users/Ben/Development/test_app/node_modules/node-llama-cpp/llama/llama.cpp/CMakeLists.txt
  Building Custom Rule C:/Users/Ben/Development/test_app/node_modules/node-llama-cpp/llama/llama.cpp/CMakeLists.txt
  Compiling CUDA source file ..\..\llama.cpp\ggml-cuda.cu...

  C:\Users\User\Development\test_app\node_modules\node-llama-cpp\llama\build\llama.cpp>"Y:\CUDA_12\bin\nvcc.exe"  --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64" -x cu   -I"C:\Users\User\Devel
  opment\test_app\node_modules\node-addon-api" -I"C:\Users\User\.cmake-js\node-x64\v19.9.0\include\node" -I"C:\Users\User\Development\test_app\node_modules\node-llama-cpp\llama\llama.cpp\." -IY:\CUDA_12\include -IY:\CUDA_12\include     --keep-dir x64\Release  -maxrregcount=0
    --machine 64 --compile -cudart static --generate-code=arch=compute_52,code=[compute_52,sm_52] --generate-code=arch=compute_61,code=[compute_61,sm_61] --generate-code=arch=compute_70,code=[compute_70,sm_70] /EHsc -Xcompiler="/EHsc -Ob2"   -D_WINDOWS -DNDEBUG -DNAPI_VERSION=7 -DGGML_USE_K_
  QUANTS -DGGML_USE_CUBLAS -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -D_CRT_SECURE_NO_WARNINGS -D_XOPEN_SOURCE=600 -D"CMAKE_INTDIR=\"Release\"" -D_MBCS -DWIN32 -D_WINDOWS -DNDEBUG -DNAPI_VERSION=7 -DGGML_USE_K_QUANTS -DGGML_USE_CUBLAS -DGGML_CUDA_DMMV_X=32 -DGGML_C
  UDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -D_CRT_SECURE_NO_WARNINGS -D_XOPEN_SOURCE=600 -D"CMAKE_INTDIR=\"Release\"" -Xcompiler "/EHsc /W3 /nologo /O2 /FS   /MD /GR" -Xcompiler "/Fdggml.dir\Release\ggml.pdb" -o ggml.dir\Release\ggml-cuda.obj "C:\Users\User\Development\test_app\nod
  e_modules\node-llama-cpp\llama\llama.cpp\ggml-cuda.cu"
  nvcc fatal   : A single input file is required for a non-link phase when an outputfile is specified
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\BuildCustomizations\CUDA 12.2.targets(799,9): error MSB3721: The command ""Y:\CUDA_12\bin\nvcc.exe"  --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.
29.30133\bin\HostX64\x64" -x cu   -I"C:\Users\User\Development\test_app\node_modules\node-addon-api" -I"C:\Users\User\.cmake-js\node-x64\v19.9.0\include\node" -I"C:\Users\User\Development\test_app\node_modules\node-llama-cpp\llama\llama.cpp\." -IY:\CUDA_12\include -IY:
\CUDA_12\include     --keep-dir x64\Release  -maxrregcount=0   --machine 64 --compile -cudart static --generate-code=arch=compute_52,code=[compute_52,sm_52] --generate-code=arch=compute_61,code=[compute_61,sm_61] --generate-code=arch=compute_70,code=[compute_70,sm_70] /EHsc -Xcompiler="/EHsc
 -Ob2"   -D_WINDOWS -DNDEBUG -DNAPI_VERSION=7 -DGGML_USE_K_QUANTS -DGGML_USE_CUBLAS -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -D_CRT_SECURE_NO_WARNINGS -D_XOPEN_SOURCE=600 -D"CMAKE_INTDIR=\"Release\"" -D_MBCS -DWIN32 -D_WINDOWS -DNDEBUG -DNAPI_VERSION=7 -DGGML_USE_
K_QUANTS -DGGML_USE_CUBLAS -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -D_CRT_SECURE_NO_WARNINGS -D_XOPEN_SOURCE=600 -D"CMAKE_INTDIR=\"Release\"" -Xcompiler "/EHsc /W3 /nologo /O2 /FS   /MD /GR" -Xcompiler "/Fdggml.dir\Release\ggml.pdb" -o ggml.dir\Release\ggml-cuda.
obj "C:\Users\User\Development\test_app\node_modules\node-llama-cpp\llama\llama.cpp\ggml-cuda.cu"" exited with code 1. [C:\Users\User\Development\test_app\node_modules\node-llama-cpp\llama\build\llama.cpp\ggml.vcxproj]
ERR! OMG Process terminated: 1
```

### Steps to reproduce

Run `node-llama-cpp download --cuda` on Windows 10 with CUDA v12

### My Environment

| Operating System         |     Windows 10     |
| Node.js version          | 19.9.0  |
| `node-llama-cpp` version | 2.4.0  |

### Relevant Features Used

- [ ] Metal support
- [X] CUDA support
- [ ] Grammar

### Are you willing to resolve this issue by submitting a Pull Request?

No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by giladgd** [comment]
@BennyH26 There are many environment variables that you can set during the build and runtime to configure the CUDA support of `llama.cpp`; you can find more details [here](https://github.com/ggerganov/llama.cpp#cublas).

I do not have an Nvidia GPU, so I can't test how well the CUDA support works or provide better documentation for this process by myself.

I saw that someone else had trouble with the CUDA support and found a solution to their issue, I think it can help you: https://github.com/withcatai/node-llama-cpp/issues/22#issuecomment-1712973757

Please update if you managed to solve this issue, as I would like to document these problems and solutions to make it more straightforward to set up.

**Comment by giladgd** [comment]
Closed due to inactivity, as I assume this issue was fixed as part of #37, and I've got [confirmation that CUDA works](https://github.com/withcatai/node-llama-cpp/issues/22#issuecomment-1722461305) (on Linux though).

@BennyH26 Please update whether CUDA support now works for you on Windows.

**Comment by giladgd** [comment]
@AngryBacteria Do you still experience this issue with version `2.7.2`?
If the issue still persists, please read the [CUDA guide](https://withcatai.github.io/node-llama-cpp/guide/CUDA) and let me know if it helped you or whether you still experience this issue.
